In [1]:
import pandas as pd

Problema a resolver:

Num problema de classificação binária com alto desbalanceamento de classes, os algoritmos de classificação tender ter um alto erro/bias para a classe majoritária
Ex:
Se em 1000 exemplos, 950 são positivos, e 50 negativos. Os modelos tendem a prever mais o positivo.

1) Como medir isso?
No caso do exemplo acima, o número de erros pelo class imbalance é igual ao número de falsos positivos. I.e., previu positivo mas na vdd é negativo.


In [6]:
df = pd.read_csv("./class_imbalance/dataset_1039_hiva_agnostic.csv")
df

,attr0,attr1,attr2,attr3,attr4,attr5,attr6,attr7,attr8,attr9,...,attr1608,attr1609,attr1610,attr1611,attr1612,attr1613,attr1614,attr1615,attr1616,label
0,0,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,-1
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,-1
3,0,0,0,0,0,0,0,1,0,0,...,0,0,0,0,0,0,0,0,0,-1
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,1,0,-1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4224,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
4225,0,0,0,0,0,0,0,0,0,0,...,0,0,1,0,0,0,0,0,0,-1
4226,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
4227,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1


In [28]:
df[df["label"]==-1]

,attr0,attr1,attr2,attr3,attr4,attr5,attr6,attr7,attr8,attr9,...,attr1608,attr1609,attr1610,attr1611,attr1612,attr1613,attr1614,attr1615,attr1616,label
0,0,0,0,0,0,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,-1
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,-1
3,0,0,0,0,0,0,0,1,0,0,...,0,0,0,0,0,0,0,0,0,-1
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,1,0,-1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4224,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
4225,0,0,0,0,0,0,0,0,0,0,...,0,0,1,0,0,0,0,0,0,-1
4226,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1
4227,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,-1


De 4229 linhas, apenas 149 linhas tem label = 1

Se nós tratarmos: -1 como positivo e 1 como negativo, podemos treinar o modelo. Para esse dataset nao é necessário tratar os dados

In [29]:
split_idx = int(len(df) * 0.8)

X_df = df.drop(columns=["label"])
y_df = (df["label"] == 1).astype(int)

X_train = X_df.iloc[:split_idx].values.astype(float)
X_test = X_df.iloc[split_idx:].values.astype(float)

y_train = y_df.iloc[:split_idx].values
y_test = y_df.iloc[split_idx:].values

In [30]:
from logistic_regression import gradient_descent

weight, bias = gradient_descent(X_train, y_train)

print(weight, bias)

[ 0.00159661  0.03069626  0.00236592 ... -0.00642482  0.01194878
  0.0078136 ] -0.23708595053946335


In [31]:
from logistic_regression import predict

y_pred = predict(X_test, weight, bias)
len(y_pred)

846

In [32]:
from metrics import confusion_matrix, false_positive_rate

matrix = confusion_matrix(y_test, y_pred)
print("confusion matrix [[TN, FP], [FN, TP]]")
print(matrix)
print(f"FPR: {false_positive_rate(matrix):.4f}")


array([[  0,   0],
       [ 35, 811]])

That is, we got 35 FNs in 846 predictions. That means, 35 values should be positive but were predicted as negative.

TODO: 
1. Check the metric that check the number of FPs or TNs. Does that metric will really help check class imbalance?
2. Change some datasets (~20) so it is formatted and the majority class is the positive. Then the FPs will be the wrong ones
3. Study how I can "solve" class imbalance


1. Let's model all majority class as Positive. This is the FPR actually measure the minority error.
False Positive Rate: FP / (FP + TN) -> Number of actual negatives that were incorrectly labeled as positive.
What more? how can I develop more elaborate statistics to help that 

### Dataset 1 (dataset_1039_hiva_agnostic.csv)

### Datasets Cleaning

In [181]:
df = pd.read_csv("class_imbalance/dataset_1050_pc3.csv")
df

,LOC_BLANK,BRANCH_COUNT,CALL_PAIRS,LOC_CODE_AND_COMMENT,LOC_COMMENTS,CONDITION_COUNT,CYCLOMATIC_COMPLEXITY,CYCLOMATIC_DENSITY,DECISION_COUNT,DECISION_DENSITY,...,NODE_COUNT,NORMALIZED_CYLOMATIC_COMPLEXITY,NUM_OPERANDS,NUM_OPERATORS,NUM_UNIQUE_OPERANDS,NUM_UNIQUE_OPERATORS,NUMBER_OF_LINES,PERCENT_COMMENTS,LOC_TOTAL,c
0,2,1.0,0,0,0,0.0,1.0,0.10,0.0,0.00,...,2.0,0.08,28.0,29.0,9.0,5,13.0,0.00,10.0,False
1,1,1.0,4,0,0,0.0,1.0,0.07,0.0,0.00,...,6.0,0.06,52.0,55.0,26.0,14,16.0,0.00,14.0,False
2,27,19.0,1,4,13,26.0,11.0,0.26,12.0,2.17,...,25.0,0.13,58.0,78.0,30.0,24,83.0,30.91,42.0,False
3,2,17.0,2,0,0,24.0,9.0,0.47,8.0,3.00,...,25.0,0.41,73.0,81.0,23.0,20,22.0,0.00,19.0,False
4,6,1.0,1,0,2,0.0,1.0,0.11,0.0,0.00,...,3.0,0.06,19.0,23.0,15.0,7,18.0,18.18,9.0,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1558,6,3.0,3,0,8,4.0,2.0,0.40,2.0,2.00,...,7.0,0.10,16.0,20.0,13.0,9,20.0,61.54,5.0,False
1559,3,5.0,3,0,0,6.0,3.0,0.33,2.0,3.00,...,9.0,0.23,20.0,32.0,10.0,14,13.0,0.00,9.0,False
1560,0,1.0,0,0,0,0.0,1.0,1.00,0.0,0.00,...,2.0,1.00,0.0,2.0,0.0,2,1.0,0.00,0.0,False
1561,0,7.0,0,0,0,10.0,4.0,0.36,4.0,2.50,...,9.0,0.33,19.0,23.0,13.0,16,12.0,0.00,11.0,False


In [182]:
df["c"].value_counts()

c
False    1403
True      160
Name: count, dtype: int64

In [183]:
## Rename -> target
df = df.rename(columns={"c": "target"})
df

,LOC_BLANK,BRANCH_COUNT,CALL_PAIRS,LOC_CODE_AND_COMMENT,LOC_COMMENTS,CONDITION_COUNT,CYCLOMATIC_COMPLEXITY,CYCLOMATIC_DENSITY,DECISION_COUNT,DECISION_DENSITY,...,NODE_COUNT,NORMALIZED_CYLOMATIC_COMPLEXITY,NUM_OPERANDS,NUM_OPERATORS,NUM_UNIQUE_OPERANDS,NUM_UNIQUE_OPERATORS,NUMBER_OF_LINES,PERCENT_COMMENTS,LOC_TOTAL,target
0,2,1.0,0,0,0,0.0,1.0,0.10,0.0,0.00,...,2.0,0.08,28.0,29.0,9.0,5,13.0,0.00,10.0,False
1,1,1.0,4,0,0,0.0,1.0,0.07,0.0,0.00,...,6.0,0.06,52.0,55.0,26.0,14,16.0,0.00,14.0,False
2,27,19.0,1,4,13,26.0,11.0,0.26,12.0,2.17,...,25.0,0.13,58.0,78.0,30.0,24,83.0,30.91,42.0,False
3,2,17.0,2,0,0,24.0,9.0,0.47,8.0,3.00,...,25.0,0.41,73.0,81.0,23.0,20,22.0,0.00,19.0,False
4,6,1.0,1,0,2,0.0,1.0,0.11,0.0,0.00,...,3.0,0.06,19.0,23.0,15.0,7,18.0,18.18,9.0,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1558,6,3.0,3,0,8,4.0,2.0,0.40,2.0,2.00,...,7.0,0.10,16.0,20.0,13.0,9,20.0,61.54,5.0,False
1559,3,5.0,3,0,0,6.0,3.0,0.33,2.0,3.00,...,9.0,0.23,20.0,32.0,10.0,14,13.0,0.00,9.0,False
1560,0,1.0,0,0,0,0.0,1.0,1.00,0.0,0.00,...,2.0,1.00,0.0,2.0,0.0,2,1.0,0.00,0.0,False
1561,0,7.0,0,0,0,10.0,4.0,0.36,4.0,2.50,...,9.0,0.33,19.0,23.0,13.0,16,12.0,0.00,11.0,False


In [184]:
import numpy as np
import pandas as pd

# Remove accidental index columns
df = df.loc[:, ~df.columns.str.contains("^Unnamed")]

# Normalize column names
df.columns = df.columns.str.strip()

# Drop constant columns
constant_columns = [
    col for col in df.columns
    if col != "target" and df[col].nunique(dropna=False) <= 1
]

df = df.drop(columns=constant_columns)

# Split temporarily
y_df = df["target"]
X_df = df.drop(columns=["target"])

# Fill categorical missing values before dummies
categorical_columns = X_df.select_dtypes(include=["object", "string"]).columns

for col in categorical_columns:
    X_df[col] = X_df[col].astype("string").str.strip()
    X_df[col] = X_df[col].fillna("__missing__")

# Generate dummy variables
X_df = pd.get_dummies(
    X_df,
    columns=categorical_columns,
    drop_first=True,
    dtype=float,
)

# Fill numeric missing values
X_df = X_df.fillna(X_df.mean(numeric_only=True))
X_df = X_df.fillna(0)

# Drop constant dummy/numeric columns after encoding
X_df = X_df.loc[:, X_df.nunique(dropna=False) > 1]

# Put target back as the final column
df_clean = pd.concat(
    [X_df.reset_index(drop=True), y_df.reset_index(drop=True).rename("target")],
    axis=1,
)


In [185]:
## Majority class -> 1 and Minority class -> 0
df_clean["target"] = (df_clean["target"] == False).astype(int)

df_clean

,LOC_BLANK,BRANCH_COUNT,CALL_PAIRS,LOC_CODE_AND_COMMENT,LOC_COMMENTS,CONDITION_COUNT,CYCLOMATIC_COMPLEXITY,CYCLOMATIC_DENSITY,DECISION_COUNT,DECISION_DENSITY,...,NODE_COUNT,NORMALIZED_CYLOMATIC_COMPLEXITY,NUM_OPERANDS,NUM_OPERATORS,NUM_UNIQUE_OPERANDS,NUM_UNIQUE_OPERATORS,NUMBER_OF_LINES,PERCENT_COMMENTS,LOC_TOTAL,target
0,2,1.0,0,0,0,0.0,1.0,0.10,0.0,0.00,...,2.0,0.08,28.0,29.0,9.0,5,13.0,0.00,10.0,1
1,1,1.0,4,0,0,0.0,1.0,0.07,0.0,0.00,...,6.0,0.06,52.0,55.0,26.0,14,16.0,0.00,14.0,1
2,27,19.0,1,4,13,26.0,11.0,0.26,12.0,2.17,...,25.0,0.13,58.0,78.0,30.0,24,83.0,30.91,42.0,1
3,2,17.0,2,0,0,24.0,9.0,0.47,8.0,3.00,...,25.0,0.41,73.0,81.0,23.0,20,22.0,0.00,19.0,1
4,6,1.0,1,0,2,0.0,1.0,0.11,0.0,0.00,...,3.0,0.06,19.0,23.0,15.0,7,18.0,18.18,9.0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1558,6,3.0,3,0,8,4.0,2.0,0.40,2.0,2.00,...,7.0,0.10,16.0,20.0,13.0,9,20.0,61.54,5.0,1
1559,3,5.0,3,0,0,6.0,3.0,0.33,2.0,3.00,...,9.0,0.23,20.0,32.0,10.0,14,13.0,0.00,9.0,1
1560,0,1.0,0,0,0,0.0,1.0,1.00,0.0,0.00,...,2.0,1.00,0.0,2.0,0.0,2,1.0,0.00,0.0,1
1561,0,7.0,0,0,0,10.0,4.0,0.36,4.0,2.50,...,9.0,0.33,19.0,23.0,13.0,16,12.0,0.00,11.0,1


In [186]:
df_clean["target"].value_counts()

target
1    1403
0     160
Name: count, dtype: int64

In [187]:
df_clean.to_csv("class_imbalance_refactored/df20.csv", index=False)

In [188]:
def regenerate_ipums_dataset(source_file, output_file):
    source = pd.read_csv(f"class_imbalance/{source_file}")
    source = source.loc[:, ~source.columns.str.contains("^Unnamed")]
    source.columns = source.columns.str.strip()

    y = source["binaryClass"].map({"N": 0, "P": 1})
    if y.isna().any():
        raise ValueError(f"Unexpected labels in {source_file}: {source['binaryClass'].unique()}")

    X = source.drop(columns=["binaryClass"])
    categorical_columns = X.select_dtypes(include=["object", "string"]).columns

    for col in categorical_columns:
        X[col] = X[col].astype("string").str.strip().fillna("__missing__")

    X = pd.get_dummies(X, columns=categorical_columns, drop_first=True, dtype=float)
    X = X.apply(pd.to_numeric, errors="coerce")
    X = X.fillna(X.mean(numeric_only=True)).fillna(0)
    X = X.loc[:, X.nunique(dropna=False) > 1]

    df_clean = pd.concat(
        [X.reset_index(drop=True), y.reset_index(drop=True).rename("target")],
        axis=1,
    )
    df_clean.to_csv(f"class_imbalance_refactored/{output_file}", index=False)
    return df_clean["target"].value_counts()

print("df3", regenerate_ipums_dataset("dataset_1018_ipums_la_99-small.csv", "df3.csv"))
print("df4", regenerate_ipums_dataset("dataset_1002_ipums_la_98-small.csv", "df4.csv"))


df3 target
1    8276
0     568
Name: count, dtype: int64
df4 target
1    6694
0     791
Name: count, dtype: int64
